components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/transaction/UtxoTransactionBuilderImpl.kt

```
@@ -33,26 +33,47 @@ class UtxoTransactionBuilderImpl(
     private var alreadySigned = false
 
     override fun addAttachment(attachmentId: SecureHash): UtxoTransactionBuilder {
+        require(attachmentId !in attachments) {
+            "Duplicating attachments is not allowed."
+        }
         this.attachments += attachmentId
         return this
     }
 
     override fun addCommand(command: Command): UtxoTransactionBuilder {
+        require(command !in commands) {
+            "Duplicating commands is not allowed."
+        }
         this.commands += command
         return this
     }
 
     override fun addSignatories(signatories: Iterable<PublicKey>): UtxoTransactionBuilder {
+        require(
+            this.signatories.intersect(signatories.toSet()).isEmpty()
+                    && signatories.distinct().size == signatories.count()
+        ) {
+            "Duplicating signatories is not allowed."
+        }
         this.signatories += signatories
         return this
     }
 
     override fun addInputState(stateRef: StateRef): UtxoTransactionBuilder {
+        require(stateRef !in inputStateRefs) {
+            "Duplicating input StateRefs is not allowed."
+        }
         this.inputStateRefs += stateRef
         return this
     }
 
     override fun addInputStates(stateRefs: Iterable<StateRef>): UtxoTransactionBuilder {
+        require(
+            this.inputStateRefs.intersect(stateRefs.toSet()).isEmpty()
+                    && stateRefs.distinct().size == stateRefs.count()
+        ) {
+            "Duplicating input StateRefs is not allowed."
+        }
```

Why have we decided to make everything distinct? I noticed this in the flows as well when updating the transaction builder.

I don't disagree, but worried that we added this out of no where.

Do we want to throw or should we instead only add if its distinct?

---

We discussed it here:
https://github.com/corda/corda-runtime-os/pull/3195#discussion_r1119865560
First I added only the first check, then realized that without the second part, duplications could still sneak in.

For the tx builder, I felt that better to be explicit (require/throw) than silently ignore the duplication.

---

Not sure if C4 checked this, but in practice you can only consume a state once, so you should only add a specific StateRef once at most to inputs. Also note another validity check (?is this in verifier?) is that a StateRef cannot be both in inputs and reference inputs. Either it is consumed, or it is not consumed can't be both at once

---

It was in verifier. I added it there.

---

I have reverted this as we discussed on slack.